File: practice/LeetCode/Longest_Increasing_Subsequence.java
// https://leetcode.com/problems/longest-increasing-subsequence/
// Accepted    91 ms

public class Solution {
    public int lengthOfLIS(int[] nums) {
        int memo[][] = new int[nums.length + 2][nums.length];
        for (int[] l : memo) {
            Arrays.fill(l, -1);
        }
        return lengthofLIS(nums, 0, 0, memo);
    }
    public int lengthofLIS(int[] nums, int previndex, int curpos, int[][] memo) {
        if (curpos == nums.length) {
            return 0;
        }
        if (memo[previndex][curpos] >= 0) {
            return memo[previndex][curpos];
        }
        int taken = 0;
        if (previndex <= 0 || nums[curpos] > nums[previndex - 1]) {
            taken = 1 + lengthofLIS(nums, curpos + 1, curpos + 1, memo);
        }

        int nottaken = lengthofLIS(nums, previndex, curpos + 1, memo);
        memo[previndex][curpos] = Math.max(taken, nottaken);
        return memo[previndex][curpos];
    }
}

File: practice/LeetCode/Longest Increasing Subsequence.cpp
// https://leetcode.com/problems/longest-increasing-subsequence/

class Solution {
public:
    int lengthOfLIS(vector<int>& nums, int i = 0, int j = -1) {
        // if (i == nums.size()) {
        //     return 0;
        // }
        // int ans = 0;
        // if (j == -1 || nums[i] > nums[j]) {
        //     ans = 1 + lengthOfLIS(nums, i + 1, i);
        // }
        // ans = max(ans, lengthOfLIS(nums, i + 1, j));
        // return ans;


        int n = nums.size();
        vector<int> dp(n, 1);
        dp[0] = 1;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[i], 1 + dp[j]);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};